File: analyze/convert-to-csv-real/csv-network.py
import json
import csv
import argparse
from pathlib import Path

def process_packets_splitter(input_file_path):
    """
    Reads JSON packet data and splits it into multiple CSV files.
    Designed to be 'lossless' based on the nuclear jq structure provided.
    
    Captures:
    - Standard 5-tuple (IPs, Ports, Protocol)
    - ARP Specifics (Operation, Protocol, Addresses)
    - IP Specifics (V4 flag)
    - Ethernet & DNS details (if present)
    """
    
    initialized_paths = set()

    # --- COMPREHENSIVE HEADER ---
    header = [
        'timestamp', 
        'device', 
        'length', 
        'link_protocol', 
        'network_protocol', 
        'transport_protocol', 
        'application_protocol',
        # IP Specifics
        'ip_v4',
        'src_ip', 
        'dst_ip', 
        # TCP / UDP (Ports)
        'src_port', 
        'dst_port',
        # ARP Specifics
        'arp_operation',
        'arp_protocol',
        # Ethernet (MACs) - kept for robustness
        'eth_src_mac', 
        'eth_dst_mac',
        # DNS specific - kept for robustness
        'dns_query_id', 
        'dns_query_name' 
    ]

    try:
        with open(input_file_path, 'r') as infile:
            for line in infile:
                if not line.strip(): 
                    continue
                
                try:
                    log_entry = json.loads(line)
                    if not isinstance(log_entry, dict):
                        continue
                except json.JSONDecodeError:
                    continue

                # --- 1. Routing via Outer DevID ---
                dev_id = log_entry.get('DevID', 'unknown')
                parts = dev_id.split('.')
                
                if len(parts) >= 2:
                    folder_sub = f"{parts[0]}-data"
                    folder_base = "_".join(parts[1:])
                else:
                    folder_base = "unknown_device_group"
                    folder_sub = f"{dev_id}-data"

                output_dir = Path(folder_base) / folder_sub
                output_file = output_dir / "network.csv"
                output_path_str = str(output_file)

                # --- 2. File Setup ---
                output_dir.mkdir(parents=True, exist_ok=True)
                write_header = False
                
                if output_path_str not in initialized_paths:
                    if not output_file.exists():
                        write_header = True
                    initialized_paths.add(output_path_str)

                # --- 3. Process Packets ---
                with open(output_file, 'a', newline='') as outfile:
                    writer = csv.writer(outfile)
                    
                    if write_header:
                        writer.writerow(header)

                    packets_list = log_entry.get("Packets", [])
                    if not isinstance(packets_list, list):
                        packets_list = []

                    for packet in packets_list:
                        if not isinstance(packet, dict):
                            continue

                        # --- 4. Extract Data Fields ---
                        
                        # Basic Info
                        timestamp = packet.get('TimeStamp', 0)
                        device = packet.get('Dev', 'N/A')
                        length = packet.get('Length', 0)
                        link_proto = packet.get('LinkProtocol', 'N/A')
                        net_proto = packet.get('NetworkProtocol', 'N/A')
                        trans_proto = packet.get('TransportProtocol', 'N/A')
                        app_proto = packet.get('ApplicationProtocol', 'N/A')

                        # Initialize variables for specific protocols
                        ip_v4 = 'N/A'
                        src_ip, dst_ip = 'N/A', 'N/A'
                        src_port, dst_port = 'N/A', 'N/A'
                        arp_op, arp_prot = 'N/A', 'N/A'
                        eth_src, eth_dst = 'N/A', 'N/A'
                        dns_id, dns_name = 'N/A', 'N/A'

                        # --- IP Layer Processing ---
                        if 'IP' in packet:
                            ip_data = packet['IP']
                            src_ip = ip_data.get('SRCIP', 'N/A')
                            dst_ip = ip_data.get('DSTIP', 'N/A')
                            ip_v4 = ip_data.get('V4', 'N/A') # Capture IP.V4
                            
                            # Check for Ports (UDP/TCP are usually inside IP packets)
                            if 'UDP' in packet:
                                src_port = packet['UDP'].get('SrcPort', 'N/A')
                                dst_port = packet['UDP'].get('DstPort', 'N/A')
                            elif 'TCP' in packet:
                                src_port = packet['TCP'].get('SrcPort', 'N/A')
                                dst_port = packet['TCP'].get('DstPort', 'N/A')
                        
                        # --- ARP Layer Processing ---
                        elif 'ARP' in packet:
                            arp_data = packet['ARP']
                            src_ip = arp_data.get('SrcProtAdd', 'N/A') # Mapping ARP IPs to main IP cols
                            dst_ip = arp_data.get('DstProtAdd', 'N/A')
                            arp_op = arp_data.get('Operation', 'N/A')  # Capture ARP.Operation
                            arp_prot = arp_data.get('Protocol', 'N/A') # Capture ARP.Protocol

                        # --- Ethernet Layer ---
                        if 'ETH' in packet:
                            # Safely handle ETH if it's a dict
                            if isinstance(packet['ETH'], dict):
                                eth_src = packet['ETH'].get('SrcMac', 'N/A')
                                eth_dst = packet['ETH'].get('DstMac', 'N/A')
                            # If ETH is just a string/flag (based on your jq output), we leave as N/A

                        # --- DNS Layer ---
                        if 'DNS' in packet:
                            if isinstance(packet['DNS'], dict):
                                dns_id = packet['DNS'].get('ID', 'N/A')
                                questions = packet['DNS'].get('Query', [])
                                if isinstance(questions, list) and len(questions) > 0:
                                    dns_name = questions[0].get('Name', 'N/A')
                                elif isinstance(questions, dict):
                                    dns_name = questions.get('Name', 'N/A')

                        # Construct Row
                        row = [
                            timestamp, 
                            device, 
                            length, 
                            link_proto, 
                            net_proto, 
                            trans_proto, 
                            app_proto,
                            ip_v4,          # New
                            src_ip, 
                            dst_ip, 
                            src_port, 
                            dst_port,
                            arp_op,         # New
                            arp_prot,       # New
                            eth_src, 
                            eth_dst,
                            dns_id, 
                            dns_name
                        ]
                        
                        writer.writerow(row)

    except FileNotFoundError:
        print(f"Error: Input file not found at {input_file_path}")
    except IOError as e:
        print(f"File I/O Error: {e}")
    else:
        print(f"Processing complete. Data distributed into {len(initialized_paths)} unique CSV files.")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Splits Network Packet JSON logs (Lossless based on nuclear jq scan).",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    parser.add_argument("input_file", type=Path, help="Path to the input file.")
    args = parser.parse_args()
    process_packets_splitter(args.input_file)

File: analyze/convert-to-csv-real/csv-network.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <filesystem>
#include <sstream>
#include <set>
#include <unordered_map>
#include "json.hpp" // https://github.com/nlohmann/json

using json = nlohmann::json;
namespace fs = std::filesystem;


/* 

This program reads JSON interface data and splits it into multiple CSV files 
based on the DevID structure.
This program require the json hpp file to compile: https://github.com/nlohmann/json

Path Structure: ./[suffix_path]/[prefix]-data/network.csv
Example: client.a.b.c.d -> ./a_b_c_d/client-data/network.csv

*/


// --- CONFIGURATION ---
const int BATCH_SIZE = 50000; // Buffer 50000 packets in RAM before writing to disk

// --- Helper: Safely extract values ---
std::string get_val(const json& j, const std::string& key, const std::string& default_val = "N/A") {
    if (!j.contains(key) || j[key].is_null()) return default_val;
    const auto& val = j[key];
    
    if (val.is_string()) return val.get<std::string>();
    if (val.is_number_integer() || val.is_number_unsigned()) return std::to_string(val.get<int64_t>());
    if (val.is_number_float()) return std::to_string(val.get<double>());
    if (val.is_boolean()) return val.get<bool>() ? "true" : "false";
    
    return default_val;
}

// --- Helper: Split string ---
std::vector<std::string> split(const std::string& s, char delimiter) {
    std::vector<std::string> tokens;
    std::string token;
    std::istringstream tokenStream(s);
    while (std::getline(tokenStream, token, delimiter)) {
        tokens.push_back(token);
    }
    return tokens;
}

// --- FLUSH FUNCTION (Writes RAM buffer to Disk) ---
void flush_buffers(std::unordered_map<std::string, std::string>& buffer_map, std::set<std::string>& initialized_paths) {
    for (auto& [file_path, data] : buffer_map) {
        if (data.empty()) continue;

        fs::path p(file_path);
        
        // Ensure folder exists
        if (!fs::exists(p.parent_path())) {
            fs::create_directories(p.parent_path());
        }

        // Header Logic
        bool write_header = false;
        if (initialized_paths.find(file_path) == initialized_paths.end()) {
            if (!fs::exists(p)) write_header = true;
            initialized_paths.insert(file_path);
        }

        // Write chunk to file
        std::ofstream outfile(file_path, std::ios::app);
        if (outfile.is_open()) {
            if (write_header) {
                outfile << "timestamp,device,length,link_protocol,network_protocol,"
                        << "transport_protocol,application_protocol,ip_version,src_ip,dst_ip,"
                        << "src_port,dst_port,arp_operation,arp_protocol,arp_src_proto,arp_dst_proto,"
                        << "eth_src_mac,eth_dst_mac\n";
            }
            outfile << data; 
            outfile.close();
        }
    }
    // Clear RAM for next batch
    buffer_map.clear();
}

int main(int argc, char* argv[]) {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <input_file>" << std::endl;
        return 1;
    }

    std::string input_path = argv[1];
    std::ifstream infile(input_path);
    if (!infile.is_open()) {
        std::cerr << "Error: Could not open input file " << input_path << std::endl;
        return 1;
    }

    std::set<std::string> initialized_paths;
    std::unordered_map<std::string, std::string> buffer_map; // The RAM Buffer
    
    std::string line;
    int line_count = 0;
    int processed_count = 0; // Just for visual feedback if needed

    while (std::getline(infile, line)) {
        if (line.empty()) continue;

        try {
            auto log_entry = json::parse(line, nullptr, false);
            if (log_entry.is_discarded() || !log_entry.is_object()) continue;

            // --- 1. Routing Logic ---
            std::string dev_id = get_val(log_entry, "DevID", "unknown");
            std::vector<std::string> parts = split(dev_id, '.');
            
            std::string folder_base, folder_sub;
            if (parts.size() >= 2) {
                folder_sub = parts[0] + "-data";
                folder_base = parts[1];
                for(size_t i = 2; i < parts.size(); ++i) folder_base += "_" + parts[i];
            } else {
                folder_base = "unknown_device_group";
                folder_sub = dev_id + "-data";
            }

            fs::path output_dir = fs::path(folder_base) / folder_sub;
            fs::path output_file = output_dir / "network.csv";
            std::string output_path_str = output_file.string();

            // --- 2. Extract Data (New Flat Structure) ---
            // Note: In your new data, keys are at the top level.
            
            // Core
            std::string timestamp = get_val(log_entry, "TimeStamp", "0");
            std::string device    = get_val(log_entry, "Dev");
            std::string length    = get_val(log_entry, "Length", "0");
            std::string link_p    = get_val(log_entry, "LinkProtocol");
            std::string net_p     = get_val(log_entry, "NetworkProtocol");
            std::string trans_p   = get_val(log_entry, "TransportProtocol");
            std::string app_p     = get_val(log_entry, "ApplicationProtocol");

            // IP Data (Direct keys)
            std::string ip_ver = get_val(log_entry, "Version", "N/A");
            std::string src_ip = get_val(log_entry, "SRC_IP", "N/A");
            std::string dst_ip = get_val(log_entry, "DST_IP", "N/A");

            // Ports (Direct keys)
            std::string src_port = get_val(log_entry, "SrcPort", "N/A");
            std::string dst_port = get_val(log_entry, "DstPort", "N/A");

            // ARP Data (Direct keys)
            std::string arp_op        = get_val(log_entry, "Operation", "N/A");
            std::string arp_prot      = get_val(log_entry, "Protocol", "N/A");
            std::string arp_src_proto = get_val(log_entry, "SrcProtAdd", "N/A");
            std::string arp_dst_proto = get_val(log_entry, "DstProtAdd", "N/A");

            // Ethernet Data (Direct keys)
            std::string eth_src = get_val(log_entry, "SRC_MAC", "N/A");
            std::string eth_dst = get_val(log_entry, "DST_MAC", "N/A");

            // --- 3. Append to Buffer ---
            std::string& file_buffer = buffer_map[output_path_str]; 
            
            file_buffer += timestamp + "," + device + "," + length + "," 
                         + link_p + "," + net_p + "," + trans_p + "," 
                         + app_p + "," + ip_ver + "," + src_ip + "," + dst_ip + ","
                         + src_port + "," + dst_port + "," + arp_op + "," + arp_prot + "," 
                         + arp_src_proto + "," + arp_dst_proto + ","
                         + eth_src + "," + eth_dst + "\n";

            line_count++;
            processed_count++;

            // --- 4. Check Flush Condition ---
            if (line_count >= BATCH_SIZE) {
                flush_buffers(buffer_map, initialized_paths);
                line_count = 0;
            }

        } catch (...) { continue; }
    }

    // --- 5. Final Flush ---
    flush_buffers(buffer_map, initialized_paths);

    return 0;
}
